README.md
# nebesaTv7Android

## Similar app on the other TV platforms
  - [Samsung](https://github.com/heaven-dev/tv7SamsungNg)
  - [LG](https://github.com/heaven-dev/tv7LgSpa)

## Demo

Similar app running on Samsung TV emulator (API version 6.0).

![Demo](https://github.com/heaven-dev/demo/blob/main/demo/demo.gif)

## Overview

[TV7](https://www.nebesatv7.com/) Android smart TV application. This application locale is __ru__ (Russian). Application make possible to watch TV channel and videos from the video archive.

This application is already in the [Google Play store](https://play.google.com/store/apps/details?id=ru.tv7.nebesa). There are some screenshots of the application.

Developer pages of [Android TV](https://developer.android.com/tv).
Guide of [Android TV](https://developer.android.com/training/tv).

## Instructions

### Download and install git
  - If your computer OS is windows you can download the __git__ from [here](https://git-scm.com/download/win).
  - If your computer OS is linux (Ubuntu) you can install __git__ using package manager.

### Clone repository
  - Clone this repository to your computer disk.
    - __git clone https://github.com/heaven-dev/nebesaTv7Android.git__

### Download and install Android Studio
  - You can download Android Studio from [here](https://developer.android.com/studio).
  - When the download is complete start the setup program.
    - If asked during the setup you have to install also an emulator.
    - If asked during the setup you have to select __Android 10.0 (Q) platform__.

### Open project to Android Studio
  - When the setup is complete start the Android Studio.
  - Select __Open existing Android Studio project__ to open the project.

### Install emulator
  - You can install the emulator and an other tools from here: __Tools -> SDK Manager -> Appearance & Behaviour -> System settings --> Android SDK__
    - Select __SDK Tools__ tab.
      - Select the __Android SDK Build-Tools__ if needed.
      - Select the __Android Emulator__ if needed.
      - Select the __Android SDK Platform-Tools__ if needed.
      - Select the __Google Web Driver__ if needed (probably this is needed to run the application on a real device).
      - Select __Apply__ and __OK__.

### Create virtual device
  - Select __Tools -> AVD Manager__.
    - Select __+ Create virtual device__ from the left bottom of the modal.
    - Select __TV Category__.
    - Select __Android TV (720p)__.
      - Later you can try also other resolutions.
    - Select __Next__.
    - You can try API levels from __23 to 28__, but you can select only one at this time.
    - Select __Marsmallow__ system image at first.
    - Select __Next__ and it downloads the system image if not already downloaded.
    - You can give a name to the __AVD__.
    - Select __Automatic__ from the __Graphics__ dropdown.
    - Select __Landscape__
    - Select __Finish__.
    - Your virtual device is added to the list of __Your Virtual Devices__.

### Start Virtual Device
  - Select __Tools -> AVD Manager__ if needed.
  - Start virtual device from the __green triangle__ of the __Actions__ menu.
  - You can open the menu from the __grey triangle__ of the __Actions__ menu.
    - Sometimes you need to __Wipe data__ from the virtual device if everything isn't work like expected.

### Launch application to the Virtual Device
  - Select __Run --> Run app__ to launch a release version of the application.
  - Select __Run --> Debug app__ to debug the application.
  - Select __Run --> Profile app__ to profile the application.

### Remote control of TV Emulator
  - I didn't find nice remote control for the TV emulator, but you can use simple keypad.
    - Click __three dots__ from the toolbar on the right side of emulator.
    - __Extended controls__ modal opens.
    - Select __Directional pad__ from the left side menu of the modal and you can simulate some of the remote control buttons.
    - Back button is the __triangle__ on toolbar above the __three dots__ mentioned above.

### Run application on real device
  - Follow instructions from [here](https://developer.android.com/training/tv/start/start#run-on-a-real-device).
  - [This](https://developer.android.com/studio/run/device) probably helps.

### Build project
  - Follow instructions from [here](https://developer.android.com/studio/run#reference).

### Minimum Android version to run this application
  - Lollipop
    - API level 22
    - Android 5.1 (Android TV)

### License
 - [MIT](https://github.com/heaven-dev/nebesaTv7Android/blob/master/LICENSE.md)
 - __Unauthorized publication of programs or subtitles is prohibited.__



CHANGELOG.md
# 20.07.2021 11:12

### Bug Fixes

* none

### Features

* Configuration clear functionality added to the information page.
* Added dynamic category rows to archive main page.
* New version number 23.


# 10.07.2021 16:15

### Bug Fixes

* none

### Features

* Added series programs to the archive main page.
* Added series page which can be used to add/remove series to favorites.
* New version number 22.


# 09.07.2021 08:53

### Bug Fixes
* none

### Features
* Search string appended to the title of search result page.
* Search history is available on the search page.
* Use mobile API instead of exodus API to get guide.
* New version 21.


# 30.06.2021 11:43

### Bug Fixes
* none

### Features
* Added restart button to the error page.
* Removed app built time from the about fragment.
* Changed the date format to d.m.yyyy on guides.
* New version 20.


# 10.06.2021 12:31

### Bug Fixes
* Fixed TV guide adapter bug.
* minSdkVersion changed to 22 in the AndroidManifest.xml and the build.gradle files.
* Changed date item selected/focused functionality on the guide fragment.
* Fixed TV guide fragment date and program item focus bug.
* minifyEnabled and shrinkResources set to false on the build.gradle file.

### Features
* New version 19.


# 30.05.2021 19:20

### Bug Fixes
* Fixed program title bug on search result page.
* Added app build timestamp to the info page of app.
* Fixed exit fragment bug.

### Features
* New version 18.


# 17.05.2021 15:28

### Bug Fixes
* none

### Features
* Added Android version to app info page.
* Updated the value of targetSdkVersion on build.gradle.
* Updated version name of app to 1.0.0.
* New version 17.


# 26.04.2021 15:10

### Bug Fixes
* Fixed remote home screen button related bug to archive player fragment.

### Features
* New version 16.


# 25.04.2021 13:10

### Bug Fixes
* none

### Features
* Changes to screen saver enable/disable functionality.
* New version 15.


# 20.04.2021 09:34

### Bug Fixes
* Fixed grid views refresh bug.
* Fixed pause/play bug on archive player fragment.

### Features
* Changed progress bar color to TV7 default.
* New version 14.


# 15.04.2021 08:54

### Bug Fixes
* Fixed grid views refresh bug.

### Features
* none


# 12.04.2021 10:00

### Bug Fixes
* none

### Features
* Changed the style of video subtitles.


# 06.04.2021 08:57

### Bug Fixes
* none

### Features
* Changed the style and the position of video subtitles.
* Added possibility to open another video on archive player during playback.
* Sidebar menu absolute positioning.
* Version 12.


# 19.03.2021 08:46

### Bug Fixes

* Fix to localized texts.
* Added id=null check to image URLs.

### Features

* Version 11.


# 17.03.2021 08:52

### Bug Fixes

* Fallback image changed and it was missing in some fragments.

### Features

* Version 10.

# 16.03.2021 08:07

### Bug Fixes

* none

### Features

* Added info page of channel.
* Added copyright text.
* Version 9.


# 14.03.2021 16:07

### Bug Fixes

* none

### Features

* Added stream error retry policy.
* Version 8.


# 14.03.2021 09:47

### Bug Fixes

* none

### Features

* Volley response error handling.


# 10.03.2021 13:04

### Bug Fixes

* Fix to search text box.
* Fix to virtual keyboard.

### Features

* Smaller toolbar and sidebar.
* Added video watch status bar to the program info page.
* Added caption text to controls of archive player page.
* Invalid search set text box background to red.
* Error page and new error handling.
* New version number.


# 06.02.2021 12:20

### Bug Fixes

* Fixed guide bug on TV main page.
* Fixed category, favorites, guide and search result caption texts bug.
* Version number updated.

### Features

* none


# 29.01.2021 08:33

### Bug Fixes

* Fixed progressbar bug in archive player view.

### Features

* Category item logo changed to channel logo on archive main page.
* Border of keys of virtual keyboard changed.
* Prevent a screensaver to appear when the application is running.
* App label changed to Небеса ТВ7.
* Smooth vertical scroll to archive main page.
* Version number to 5.


# 19.01.2021 16:10

### Bug Fixes

* none

### Features

* Added info about aspect ratio to program info page if aspect ratio isn't 16:9.
* Order of recommendations methods call changed in the archive main fragment.


# 06.01.2021 18:02

### Bug Fixes

* Added handling to the KEYCODE_MEDIA_PLAY_PAUSE keycode event to video fragments. Version number increased.

### Features

* none


# 03.01.2021 09:45

### Bug Fixes

* If video stream of the video of video archive is active and the HOME button of the remote control is pressed the application pauses the video before
  returning the home screen of the TV.

### Features

* none


# 25.12.2020 11:17

### Bug Fixes

* When home button of the remote control is pressed and video is playing the application stops the video and release the video player. That is why the NOW
  PLAYING of the application is not shown on the home screen of the TV.
* Added support to pause and play buttons of the remote control.
* Added DPAD buttons functionality to the video views.

### Features

* none


# 17.12.2020 18:34

### Bug Fixes

* none

### Features

* App package name changed from ru.tv7.nebesatv7 to ru.tv7.nebesa.


# 16.12.2020 11:56

### Bug Fixes

* none

### Features

* About menu item of sidebar moved to bottom.


# 15.12.2020 14:06

### Bug Fixes

* none

### Features

* Added app and platform info page.


# 12.12.2020 17:48

### Bug Fixes

* Banner image changed.

### Features

* none


# 11.12.2020 08:38

### Bug Fixes

* Added localized texts related to favorites and search.

### Features

* none


# 10.12.2020 10:15

### Bug Fixes

* Fixed size of one banner.

### Features

* none


# 10.12.2020 08:30

### Bug Fixes

* App icons and banners fixed. Added own icons and banners to each resolution.

### Features

* none


# 8.12.2020 11:00

### Bug Fixes


app/src/main/java/ru/tv7/nebesa/helpers/Constants.java
package ru.tv7.nebesa.helpers;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import ru.tv7.nebesa.R;

/**
 * Constants.
 */
abstract public class Constants {
    public static final String LOG_TAG = "tv7";

    public static final boolean VOLLEY_CACHE = false;
    public static final int VOLLEY_TIMEOUT_VALUE = 30000;

    public static final int SOMETHING_WENT_WRONG_ERROR = 100;
    public static final int NO_NETWORK_CONNECTION_ERROR = 101;
    public static final int NETWORK_REQUEST_FAILED_ERROR = 102;
    public static final int NETWORK_REQUEST_TIMEOUT_ERROR = 103;

    public static final int STREAM_ERROR_RETRY_DELAY = 3000;

    public static final String STREAM_URL = "https://vod.tv7.fi:443/tv7-ru/_definst_/smil:tv7-ru.smil/playlist.m3u8";
    public static final String ARCHIVE_BASE_URL = "https://www.nebesatv7.com/wp-json/tv7-api/v1/";
    public static final String ARCHIVE_VIDEO_URL = "https://vod.tv7.fi:443/vod/_definst_/mp4:_LINK_PATH_/playlist.m3u8";

    public static final String ARCHIVE_LANGUAGE = "RU1";
    public static final String TRANSLATION_LANG_ID = "ru";

    public static final String SUBTITLES_URL = "https://vod.tv7.fi:4430/vod/";
    public static final String SEARCH_URL = "https://edom.tv7.fi:8443/search1.2/SearchServlet";

    public static final float PROGRESS_BAR_SIZE = 0.85f;
    public static final int FADE_IN_ANIMATION_DURATION = 900;
    public static final float FADE_IN_ANIMATION_START = 0.0f;
    public static final float FADE_IN_ANIMATION_END = 1.0f;

    public static final int ONGOING_PROGRAM_ANIMATION_DURATION = 2000;

    public static final int FAVORITES_TEXT_ANIMATION_DURATION = 1500;
    public static final float FAVORITES_TEXT_ANIMATION_START = 1.0f;
    public static final float FAVORITES_TEXT_ANIMATION_END = 0.0f;
    public static final long FAVORITES_TEXT_ANIMATION_START_OFFSET = 500;

    public static final int PAUSE_START_ICON_ANIMATION_DURATION = 1100;
    public static final float PAUSE_START_ICON_ANIMATION_START = 1.0f;
    public static final float PAUSE_START_ICON_ANIMATION_END = 0.0f;
    public static final long PAUSE_START_ICON_ANIMATION_START_OFFSET = 250;

    public static final String SCROLL_Y = "scrollY";
    public static final long ARCHIVE_MAIN_SCROLL_Y_DURATION = 150;

    public static final String TV_MAIN_FRAGMENT = "tv_main_fragment";
    public static final String ARCHIVE_MAIN_FRAGMENT = "archive_main_fragment";
    public static final String TV_PLAYER_FRAGMENT = "tv_player_fragment";
    public static final String ARCHIVE_PLAYER_FRAGMENT = "archive_player_fragment";
    public static final String CATEGORIES_FRAGMENT = "categories_fragment";
    public static final String SERIES_FRAGMENT = "series_fragment";
    public static final String GUIDE_FRAGMENT = "guide_fragment";
    public static final String PROGRAM_INFO_FRAGMENT = "program_info_fragment";
    public static final String SERIES_INFO_FRAGMENT = "series_info_fragment";
    public static final String SEARCH_FRAGMENT = "search_fragment";
    public static final String SEARCH_RESULT_FRAGMENT = "search_result_fragment";
    public static final String FAVORITES_FRAGMENT = "favorites_fragment";
    public static final String CHANNEL_INFO_FRAGMENT = "channel_info_fragment";
    public static final String ABOUT_FRAGMENT = "about_fragment";
    public static final String ERROR_FRAGMENT = "error_fragment";
    public static final String EXIT_OVERLAY_FRAGMENT = "exit_overlay_fragment";

    public static final String CHANNEL_URL_PARAM = "channel_url";
    public static final String _LINK_PATH_ = "_LINK_PATH_";

    public static final boolean SHOW_ANIMATIONS = true;

    public static final int PROGRAM_VISIBLE_IMAGE_COUNT = 11;
    public static final int GUIDE_ELEMENT_COUNT = 10;
    public static final int GUIDE_TIMER_TIMEOUT = 10000;
    public static final int VIDEO_POSITION_TIMEOUT = 1000;
    public static final int VIDEO_CONTROLS_TIMEOUT = 6000;

    public static final float CUE_LINE_POSITION = 0.82f;
    public static final int SUBTITLE_BACKGROUND = 0x99000000;

    public static final int DATES_COUNT = 7;
    public static final int ARCHIVE_MAIN_NO_SEL_POS = -1;
    public static final int NO_MORE_PAGING_DATA = -1;

    public static final int PROGRAM_LIST_MIN_SIZE = 22;

    public static final String PNID_PARAM = "pnid";
    public static final String AUDIO_INDEX_PARAM = "audioindex";

    public static final String AUDIO_INDEX_ENABLE_LANG = "RU1";

    public static final String TV_BRAND = "AndroidTV";

    public static final String OS_VERSION = "os.version";
    public static final String LEFT_BRACKET = "(";
    public static final String RIGHT_BRACKET = ")";

    public static final String QUESTION_MARK = "?";
    public static final String AMPERSAND = "&";
    public static final String EQUAL = "=";
    public static final String SPACE = " ";
    public static final String EMPTY = "";
    public static final String PIPE_WITH_SPACES = " | ";
    public static final String DASH = "-";
    public static final String DASH_WITH_SPACES = " - ";
    public static final String SLASH_WITH_SPACES = " / ";
    public static final String COLON = ":";
    public static final String COLON_WITH_SPACE = ": ";
    public static final String DOT = ".";
    public static final String ZERO_STR = "0";
    public static final String NEGATIVE_ONE_STR = "-1";
    public static final String UTC = "UTC";
    public static final String HTTP = "http";
    public static final String HTTPS = "https";
    public static final String UTF_8 = "UTF-8";
    public static final String ID_NULL = "id=null";

    public static final String ONE_STR = "1";
    public static final String TWO_STR = "2";
    public static final String POSITION = "position";
    public static final String PERCENT = "percent";

    public static final String FAVORITES_SP_TAG = "favorites_shared_pref";
    public static final String FAVORITES_SP_DEFAULT = "[]";

    public static final String VIDEO_STATUSES_SP_TAG = "video_statuses_shared_pref";
    public static final String VIDEO_STATUSES_SP_DEFAULT = "[]";

    public static final String SAVED_SEARCH_SP_TAG = "saved_search_shared_pref";
    public static final String SAVED_SEARCH_SP_DEFAULT = "[]";
    public static final int SAVED_SEARCH_MAX_COUNT = 12;

    public static final int TOOLBAR_HEIGHT = 60;
    public static final int ARCHIVE_MAIN_TITLE_HEIGHT = 30;

    public static final String GET_= "get_";
    public static final String RECOMMENDATIONS_METHOD = "tv7_vod_recommendations";
    public static final String BROADCAST_RECOMMENDATIONS_METHOD = "tv7_broadcast_recommendations";
    public static final String MOST_VIEWED_METHOD = "get_tv7_vod_previousweek_stats";
    public static final String NEWEST_METHOD = "tv7_vod_new";
    public static final String SERIES_METHOD = "tv7_series";
    public static final String CATEGORY_ROW_ONE_METHOD = "tv7_category_row_1";
    public static final String CATEGORY_ROW_TWO_METHOD = "tv7_category_row_2";
    public static final String CATEGORY_ROW_THREE_METHOD = "tv7_category_row_3";
    public static final String CATEGORY_ROW_FOUR_METHOD = "tv7_category_row_4";
    public static final String PARENT_CATEGORIES_METHOD = "tv7_parent_categories";
    public static final String SUB_CATEGORIES_METHOD = "tv7_sub_categories";
    public static final String TRANSLATION_METHOD = "tv7_translation";
    public static final String CATEGORY_PROGRAMS_METHOD = "tv7_category_programs";
    public static final String PROGRAM_INFO_METHOD = "tv7_program_info";
    public static final String SERIES_INFO_METHOD = "tv7_series_info";
    public static final String SERIES_PROGRAMS_METHOD = "tv7_series_programs";
    public static final String GUIDE_DATE_METHOD = "tv7_tv_guide_date";
    public static final String SEARCH_METHOD = "search";

    public static final String ZERO_DATE_TIME = "00.00.0000 00:00";
    public static final String ZERO_DATE = "00.00.0000";
    public static final String ZERO_DURATION = "00:00:00";
    public static final String TIME_STAMP_FORMAT = "%02d:%02d:%02d";

    public static final String DATE_PARAM = "date";
    public static final String LIMIT_PARAM = "limit";
    public static final String OFFSET_PARAM = "offset";
    public static final String VOD_PARAM = "vod";
    public static final String PROGRAM_ID_PARAM = "program_id";
    public static final String CATEGORY_ID_PARAM = "category_id";
    public static final String CATEGORY_PARAM = "category";
    public static final String SERIES_ID_PARAM = "series_id";
    public static final String QUERY_PARAM = "query";

    public static final int RECOMMENDED_PROGRAMS_LIMIT = 30;
    public static final int NEWEST_LIMIT = 30;
    public static final int CATEGORY_ROW_LIMIT = 5;
    public static final int CATEGORY_PROGRAMS_SEARCH_LIMIT = 30;
    public static final int SERIES_PROGRAMS_SEARCH_LIMIT = 30;

    public static final String NULL_VALUE = "null";
    public static final String ID = "id";
    public static final String IMAGE_PATH = "image_path";
    public static final String LINK_PATH = "link_path";
    public static final String EPISODE_NUMBER = "episode_number";
    public static final String SID = "sid";
    public static final String CID = "cid";
    public static final String CNAME = "cname";
    public static final String CATEGORY = "category";
    public static final String SERIES_ID = "series_id";
    public static final String FIRST_BROADCAST = "first_broadcast";
    public static final String START_DATE = "start_date";
    public static final String END_DATE = "end_date";
    public static final String BROADCAST_DATE_TIME = "broadcast_date_time";
    public static final String BROADCAST_DATE = "broadcast_date";
    public static final String DURATION = "duration";
    public static final String NAME = "name";
    public static final String SERIES_NAME = "series_name";
    public static final String SERIES = "series";
    public static final String SNAME = "sname";
    public static final String SERIES_AND_NAME = "series_and_name";
    public static final String CAPTION = "caption";
    public static final String DESCRIPTION = "description";
    public static final String IS_VISIBLE_ON_VOD = "is_visible_on_vod";
    public static final String VISIBLE_ON_VOD_SINCE = "visible_on_vod_since";
    public static final String TYPE = "type";
    public static final String PATH = "path";
    public static final String PLAY = "play";
    public static final String LANG_ID = "lang_id";
    public static final String IS_SUBTITLE = "is_subtitle";
    public static final String PARENT_ID = "parent_id";
    public static final String PARENT_NAME = "parent_name";
    public static final String CATEGORY_ID = "category_id";
    public static final String CATEGORY_NAME = "category_name";
    public static final String TIME = "time";
    public static final String END_TIME = "end_time";
    public static final String FORMATTED_START_TIME = "formatted_start_time";
    public static final String FORMATTED_END_TIME = "formatted_end_time";
    public static final String START_END_TIME = "start_end_time";
    public static final String ONGOING_PROGRAM = "ongoing_program";
    public static final String ONGOING_PROGRAM_INDEX = "ongoing_program_index";
    public static final String DATE_INDEX = "date_index";
    public static final String GUIDE_DATA = "guide_data";
    public static final String BACK_TEXT = "back_text";
    public static final String CATEGORY_MORE_BOX = "category_more_box";
    public static final String RESULTS = "results";
    public static final String ASPECT_RATIO = "aspect_ratio";
    public static final String ASPECT_RATIO_16_9 = "16:9";
    public static final String IS_SERIES = "is_series";

    public static final long CACHE_EXPIRATION_TIME = 1800000;

    public static final int ARCHIVE_MAIN_ROW_COUNT = 5;
    public static final int VIDEO_SEEK_STEP_SECONDS = 10;

    public static final int RECOMMENDATIONS_ROW_ID = 0;
    public static final int MOST_VIEWED_ROW_ID = 1;
    public static final int NEWEST_ROW_ID = 2;
    public static final int CATEGORIES_ROW_ID = 3;
    public static final int SERIES_ROW_ID = 4;
    public static final int CATEGORY_ROW_ONE_ID = 5;
    public static final int CATEGORY_ROW_TWO_ID = 6;
    public static final int CATEGORY_ROW_THREE_ID = 7;
    public static final int CATEGORY_ROW_FOUR_ID = 8;

    public static final int CATEGORY_ROW_ONE = 1;
    public static final int CATEGORY_ROW_TWO = 2;
    public static final int CATEGORY_ROW_THREE = 3;
    public static final int CATEGORY_ROW_FOUR = 4;

    public static final int CATEGORY_ROW_COUNT = 4;

    public static final int DATE_INDEX_DAY_BEFORE_YESTERDAY = -2;
    public static final int DATE_INDEX_YESTERDAY = -1;
    public static final int DATE_INDEX_TODAY = 0;
    public static final int DATE_INDEX_TOMORROW = 1;

    public static final float CATEGORY_IMAGE_SIZE_IN_PERCENT = 0.6f;

    public static final List<ComingProgramImageAndTextId> COMING_PROGRAM_IMAGE_AND_TEXT = new ArrayList<>(Arrays.asList(
        new ComingProgramImageAndTextId(R.id.image1, R.id.image1Text),
        new ComingProgramImageAndTextId(R.id.image2, R.id.image2Text),
        new ComingProgramImageAndTextId(R.id.image3, R.id.image3Text),
        new ComingProgramImageAndTextId(R.id.image4, R.id.image4Text),
        new ComingProgramImageAndTextId(R.id.image5, R.id.image5Text),
        new ComingProgramImageAndTextId(R.id.image6, R.id.image6Text),
        new ComingProgramImageAndTextId(R.id.image7, R.id.image7Text),
        new ComingProgramImageAndTextId(R.id.image8, R.id.image8Text),
        new ComingProgramImageAndTextId(R.id.image9, R.id.image9Text),
        new ComingProgramImageAndTextId(R.id.image10, R.id.image10Text)
    ));

    public static final List<GuideRowId> GUIDE_ROWS = new ArrayList<>(Arrays.asList(
        new GuideRowId(R.id.row1, R.id.row1Time, R.id.row1Title, R.id.row1Desc),
        new GuideRowId(R.id.row2, R.id.row2Time, R.id.row2Title, R.id.row2Desc),
        new GuideRowId(R.id.row3, R.id.row3Time, R.id.row3Title, R.id.row3Desc),
        new GuideRowId(R.id.row4, R.id.row4Time, R.id.row4Title, R.id.row4Desc),
        new GuideRowId(R.id.row5, R.id.row5Time, R.id.row5Title, R.id.row5Desc),
        new GuideRowId(R.id.row6, R.id.row6Time, R.id.row6Title, R.id.row6Desc),
        new GuideRowId(R.id.row7, R.id.row7Time, R.id.row7Title, R.id.row7Desc),
        new GuideRowId(R.id.row8, R.id.row8Time, R.id.row8Title, R.id.row8Desc),
        new GuideRowId(R.id.row9, R.id.row9Time, R.id.row9Title, R.id.row9Desc),
        new GuideRowId(R.id.row10, R.id.row10Time, R.id.row10Title, R.id.row10Desc)
    ));

    public static final int SIDEBAR_MENU_ICON_WIDTH = 38;

    public static final List<MenuItem> MENU_ITEM_IDS = new ArrayList<>(Arrays.asList(
        new MenuItem(R.id.tvMenuContainer, R.id.tvMenuText),
        new MenuItem(R.id.archiveMenuContainer, R.id.archiveMenuText),
        new MenuItem(R.id.guideMenuContainer, R.id.guideMenuText),
        new MenuItem(R.id.searchMenuContainer, R.id.searchMenuText),
        new MenuItem(R.id.favoritesMenuContainer, R.id.favoritesMenuText),
        new MenuItem(R.id.channelInfoMenuContainer, R.id.channelInfoMenuText),
        new MenuItem(R.id.aboutMenuContainer, R.id.aboutMenuText)
    ));

    public static final List<Integer> ARCHIVE_MAIN_CONTENT_ROW_IDS = new ArrayList<>(Arrays.asList(
        R.id.recommendContainer,
        R.id.mostViewedContainer,
        R.id.newestContainer,
        R.id.categoriesContainer,
        R.id.topicalSeriesContainer,
        R.id.categoryRowOneContainer,
        R.id.categoryRowTwoContainer,
        R.id.categoryRowThreeContainer,
        R.id.categoryRowFourContainer
    ));

    public static final List<Integer> GUIDE_DATE_IDS = new ArrayList<>(Arrays.asList(
        R.id.date_0,
        R.id.date_1,
        R.id.date_2,
        R.id.date_3,
        R.id.date_4,
        R.id.date_5,
        R.id.date_6
    ));

    public static final List<Integer> SAVED_SEARCH_IDS = new ArrayList<>(Arrays.asList(
        R.id.ss_0,
        R.id.ss_1,
        R.id.ss_2,
        R.id.ss_3,
        R.id.ss_4,
        R.id.ss_5,
        R.id.ss_6,
        R.id.ss_7,
        R.id.ss_8,
        R.id.ss_9,
        R.id.ss_10,
        R.id.ss_11
    ));

    public static final List<Integer> CLEAR_MENU_IDS = new ArrayList<>(Arrays.asList(
        R.id.cm_0,
        R.id.cm_1,
        R.id.cm_2
    ));

    public static final int SPACE_BUTTON = 1;
    public static final int BACKSPACE_BUTTON = 2;
    public static final int UPPERCASE_BUTTON = 3;
    public static final int SPECIAL_CHARS_BUTTON = 4;

    public static final Map<Integer , Integer> SEARCH_TEXT_MODE_BUTTONS = new HashMap<Integer , Integer>() {{
        put(R.id.k_2_8, SPACE_BUTTON);
        put(R.id.k_2_9, BACKSPACE_BUTTON);
        put(R.id.k_2_10, UPPERCASE_BUTTON);
        put(R.id.k_2_11, SPECIAL_CHARS_BUTTON);
    }};

    public static final Map<Integer , Integer> SEARCH_SPECIAL_CHAR_MODE_BUTTONS = new HashMap<Integer , Integer>() {{
        put(R.id.k_2_8, SPACE_BUTTON);
        put(R.id.k_2_9, BACKSPACE_BUTTON);
        put(R.id.k_2_10, SPECIAL_CHARS_BUTTON);
    }};

    public static final List<Integer> CATEGORY_ROW_IDS = new ArrayList<>(Arrays.asList(
        26,
        1,
        6,
        3,
        10,
        35,
        25,
        8,
        15,
        4,
        14
    ));
}
